README.md

```diff
@@ -223,6 +223,21 @@ practice, one of the two branches (the one that was pushed to) already has the
 commit, so you would expect to see "Everything up-to-date" in one of the two
 pushes.
 
+#### Avoiding long checkout times
+
+For large repositories, the checkout can take a long time, and waste a lot of
+bandwidth. To avoid this, you can use the "partial clone" feature by replacing
+the steps with this:
+
+```yaml
+      - name: Partial clone
+        run: git clone --bare --depth=1 --single-branch --filter=blob:none ${{ github.event.repository.html_url }} .
+      - name: Push
+        run: |
+          git config http.https://github.com/.extraheader "Authorization: Basic $(echo -n x-access-token:${{ github.token }} | base64 --wrap=0)"
+          git push origin HEAD:master HEAD:main
```

Comment:
```suggestion
- name: Configure Git
  run: git config http.https://github.com/.extraheader "Authorization: Basic $(echo -n x-access-token:${{ github.token }} | base64 --wrap=0)"
- name: Partial Clone
  run: git clone --bare --depth=1 --single-branch --filter=blob:none ${{ github.event.repository.html_url }} .
- name: Push
  run: git push origin HEAD:master HEAD:main
```

Small suggestion for readability (I think it does the same thing?)

This is cool, I didn't realize you can push to multiple branches at once. If it's convenient for you, can you update the main workflow to do the same as well? If not, no worries – just commit or discard my suggestion from the UI and I can update the main one later.

Can you confirm that `--depth=1` (and `--single-branch`) will work when the push contains multiple commits? 

Admittedly my knowledge here is on very shaky ground, but I copied `fetch-depth: 0` from [an actual workflow I have in another repo](https://github.com/skylightio/skylight-docs/blob/master/.github/workflows/sync.yml). In that case, I am trying to mirror two identical repositories (an internal/private version, and the public one I linked). After some trial-and-error, I determined that `fetch-depth: 0` works.

At the time, my mental model was that I need to fetch all the commits involved the push + at least the parent of the last commit in the push, so that Git can see that the histories on the two repos are related. Otherwise, it will refuse to do the push without `--force` due to unrelated history. I would have guessed that this applies here as well, so I am a bit surprised that this (and `--single-branch`) worked.

Evidently, my mental model is not entirely correct here, so as long as you can confirm that it works then it's all good.

Reply:
> Small suggestion for readability (I think it does the same thing?)

Not quite: that `git config` call will fail because there is no Git repository (the `git clone` command will create that). I re-ordered it and force-pushed.

> Can you confirm that `--depth=1` (and `--single-branch`) will work when the push contains multiple commits?

I am 100% certain, even without testing it.

> Admittedly my knowledge here is on very shaky ground, but I copied `fetch-depth: 0` from [an actual workflow I have in another repo](https://github.com/skylightio/skylight-docs/blob/master/.github/workflows/sync.yml). In that case, I am trying to mirror two identical repositories (an internal/private version, and the public one I linked). After some trial-and-error, I determined that `fetch-depth: 0` works.

Those are mirrors between different repositories, right? In that case, Git objects will need to be pushed because they are unlikely to be present in the target repository yet. However, here we are operating with the same remote repository. We know it has all the objects because that's where we fetch from.

> At the time, my mental model was that I need to fetch all the commits involved the push + at least the parent of the last commit in the push, so that Git can see that the histories on the two repos are related. Otherwise, it will refuse to do the push without `--force` due to unrelated history. I would have guessed that this applies here as well, so I am a bit surprised that this (and `--single-branch`) worked.

Git needs to know the hash of the parent commit in the case that you want to push new objects to a repository, indeed.

But again, this does not apply here. We're trying to update a remote ref to a commit that we know is already there. You can actually see this in the output when pushing manually (in CI, the progress is suppressed because there is no interactive terminal):

```
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
[...]
    * [new branch]     HEAD:main
```

You see, it does not transfer any objects. It does not need to.

(Technical explanation: to save on bandwidth, Git determines the delta between the tip commit(s) to push and the tip commits reported by the remote repository, of _all_ refs. In other words, Git will see that the remote repository does indeed have all the objects already and make an empty pack.)

Reply:
@dscho just woke up and still trying to digest all of this - but the burning question I have is: does that mean fetch-depth: 0 is also not necessary in the main workflow? 😅 I should probably have checked at the time

Reply:
(and if that’s the case feel free to update the script and delete the explanation text for me, or I can do it later; either way this looks good and I will merge when I can computer)

Reply:
> does that mean fetch-depth: 0 is also not necessary in the main workflow?

You will be delighted to hear that it is not ;-)

> and if that’s the case feel free to update the script and delete the explanation text for me, or I can do it later; either way this looks good and I will merge when I can computer

Whoops, I think I only updated the script.

*clicketyclick*

Yep, I forgot to update the descriptions! And I also missed a few more mentions of `fetch-depth: 0`. Sorry about that. I force-pushed a fixed version (grepping for the term `fetch-depth` to make sure that I did not miss any instance).

Reply:
@dscho This makes a lot of sense to me now, thank you so much for helping me understand this and your help with updating the guide!! 🎉

Reply:
Awesome! Thank you for starting this very helpful guide. I used it to plan Git for Windows' repositories' transition.